src/impl_/pymodule.rs

```diff
@@ -52,10 +69,16 @@ impl ModuleDef {
             initializer,
         }
     }
+    /// Return module def
+    pub fn module_def(&'static self) -> *mut ffi::PyModuleDef {
+        self.ffi_def.get()
+    }
     /// Builds a module using user given initializer. Used for [`#[pymodule]`][crate::pymodule].
     pub fn make_module(&'static self, py: Python) -> PyResult<PyObject> {
         let module = unsafe {
-            Py::<PyModule>::from_owned_ptr_or_err(py, ffi::PyModule_Create(self.ffi_def.get()))?
+            let mod_def = self.ffi_def.get();
+            (*mod_def).m_slots = std::ptr::null_mut();
```

Comment:
`PyModule_Create` does not working with `m_slots` set:

```
SystemError('module foobar_module: PyModule_Create is incompatible with m_slots'), traceback: None }
```

Reply:
What would it take to change `wrap_pymodule` to use PEP 489?

Reply:
Using [PyModule_FromDefAndSpec](https://docs.python.org/3/c-api/module.html#c.PyModule_FromDefAndSpec) I guess, but I don't quite understand the `spec` argument yet.

Reply:
Unfortunately there isn't any example of `PyModule_FromDefAndSpec` usage in the docs: https://docs.python.org/3/c-api/module.html#low-level-module-creation-functions

cc @birkenfeld do you how these low level module creation functions are supposed to be used? I don't get how to obtain a `ModuleSpec` object for passing into `PyModule_FromDefAndSpec`.

Reply:
Not off the top of my head. Maybe they are supposed to be statically filled like PyTypeObjects used to be?

Reply:
It's not easy to be statically filled, [ModuleSpec](https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec) isn't a C struct but a Python class from `importlib` and it requires a name and a [loader](https://docs.python.org/3/glossary.html#term-loader) to construct.

PEP 489 also mentioned [multiple modules in one library](https://peps.python.org/pep-0489/#multiple-modules-in-one-library), but it's doing that in Python instead of using C API.

Reply:
I think I saw somewhere that the builtin `sqlite3` module now supports PEP 489. Maybe this offers some clues?

https://github.com/python/cpython/blob/f66c857572a308822c70fd25e0197b6e0dec6e34/Modules/_sqlite/module.c#L762

Reply:
Well, that's the same function we use here for the main extension module:
https://github.com/PyO3/pyo3/pull/2245/files#diff-912566621a79aeb2b0463cd8d22a412c3166600d99e1cc7f164aef67492c2429R85, it returns a module definition object instead of a module: https://docs.python.org/3/c-api/module.html#c.PyModuleDef_Init

But `wrap_pymodule!` is more about making submodules, unable to use `PyModule_Create` makes it hard to implement.

Reply:
Ah, I see. Maybe we need to think about this in combination with the syntax suggested in #694, and get away from using `wrap_pymodule!`.

Reply:
BTW `wrap_pymodule!` is also used a lot in tests, without it it's also hard to write simple module related tests without setting up proper extension modules.

Reply:
Opened a discussion: https://discuss.python.org/t/how-to-use-pymodule-fromdefandspec-c-api-to-construct-a-module/14581

Reply:
Sovled by https://discuss.python.org/t/how-to-use-pymodule-fromdefandspec-c-api-to-construct-a-module/14581/2?u=messense